### ISSUE TITLE
Missing queue family ownership transfer in compute_nbody

### ISSUE THREAD

**Issue by cforfang** [KhronosGroup/Vulkan-Samples#3]
The `compute_nbody` sample seems to be missing a queue family ownership transfer of `compute.storage_buffer` on platforms where a compute-only queue family is available.

Note how `compute.storage_buffer` is used as a storage buffer in `compute.queue` and as as vertex buffer in `queue`, and consider the following:
- If the platform has a compute only queue family available, it will be preferred.
    - The compute dispatches use the  `compute.queue` queue: https://github.com/KhronosGroup/Vulkan-Samples/blob/55abaaea0ca34288045cd52093cb1060b891508a/samples/api/compute_nbody/compute_nbody.cpp#L629
   - `compute.queue` is set here: https://github.com/KhronosGroup/Vulkan-Samples/blob/55abaaea0ca34288045cd52093cb1060b891508a/samples/api/compute_nbody/compute_nbody.cpp#L437
   - The logic in `Device::get_queue_family_index(VkQueueFlagBits queue_flag)` will pick a queue family with compute and no graphics if available; https://github.com/KhronosGroup/Vulkan-Samples/blob/55abaaea0ca34288045cd52093cb1060b891508a/framework/core/device.cpp#L349
    - **On my Radeon RX 590 queue family index 1 is picked here**
- The default `queue`'s queue family is the first one available which supports presentation:
   - It's created here https://github.com/KhronosGroup/Vulkan-Samples/blob/55abaaea0ca34288045cd52093cb1060b891508a/framework/api_vulkan_sample.cpp#L57
   - This calls `Device::get_queue_by_present` which picks the first queue family which can present; https://github.com/KhronosGroup/Vulkan-Samples/blob/55abaaea0ca34288045cd52093cb1060b891508a/framework/core/device.cpp#L328
    - **On my Radeon RX 590 queue family index 0 is picked here**
- `compute.storage_buffer` is created with `VK_SHARING_MODE_EXCLUSIVE`; see https://github.com/KhronosGroup/Vulkan-Samples/blob/55abaaea0ca34288045cd52093cb1060b891508a/framework/core/buffer.cpp#L32

In other words, `compute.queue` and `queue` can be from different queue families (and _is_ on my system), which means `compute.storage_buffer` can not be shared between the two queues without having the queue family ownership transferred:
```Resources created with a VkSharingMode of VK_SHARING_MODE_EXCLUSIVE must have their ownership explicitly transferred from one queue family to another in order to access their content in a well defined manner on a queue in a different queue family```

The problem is the sample is missing pipeline-barriers to perform this ownership transfer currently. As such, the sample is only correct on systems where the code happens to pick the queues so they are from the same queue family.

**Comment by SaschaWillems** [comment]
We'll look into this. I tested this sample on a device with separate graphics and compute queue family indices and had no problems running it.

Are you getting any validation layer messages?

**Comment by cforfang** [comment]
> Are you getting any validation layer messages?

No, nothing (using Vulkan SDK 1.1.126.0). Not that I'm surprised given synchronization validation still seems extremely lacking.

**Comment by SaschaWillems** [comment]
Can you check with this branch: https://github.com/SaschaWillems/Vulkan-Samples/tree/compute_nbody_sync

I have added proper acquire and release barriers on both queue families (if graphics and compute differ).

Note that this will trigger one initial validation message about a missing release barrier, though I'm not sure if that's my fault or if validation layers don't catch it.

Can you test if this version works fine for your setup?

**Comment by Andreyogld3d** [comment]
Hi @SaschaWillems I have integrated your fixes inside my 3DEngine, it workins fine! except only one one initial validation message that you mentioned above.

**Comment by jherico** [comment]
@SaschaWillems the intiial validation error is valid.  You create and populate the storage buffer on the primary queue in `ComputeNBody::prepare_storage_buffers`, but then the compute command buffer runs before the graphics command buffer, so there is an acquire barrier without a corresponding release barrier. 

The solution is to copy the barrier from [this line](https://github.com/SaschaWillems/Vulkan-Samples/blob/bf0c51c73d808eda268ab4585dbf126d651f98bd/samples/api/compute_nbody/compute_nbody.cpp#L147) to between the copy and flush commands at [this line](https://github.com/SaschaWillems/Vulkan-Samples/blob/bf0c51c73d808eda268ab4585dbf126d651f98bd/samples/api/compute_nbody/compute_nbody.cpp#L341).  This provides the initial release to correspond to the compute command buffer's acquire.  Subsequent releases will be provided normally by the graphics command buffer(s).

**Comment by cforfang** [comment]
@SaschaWillems Agree with @jherico's comment on the commit here: https://github.com/SaschaWillems/Vulkan-Samples/commit/bf0c51c73d808eda268ab4585dbf126d651f98bd

But one wrinkle; when waiting on `compute.semaphore` on the graphics queue, the wait is in `submit_info.pWaitDstStageMask = VK_PIPELINE_STAGE_VERTEX_INPUT_BIT` (see [here](https://github.com/SaschaWillems/Vulkan-Samples/blob/bf0c51c73d808eda268ab4585dbf126d651f98bd/samples/api/compute_nbody/compute_nbody.cpp#L687)). So for a dependency chain to form, it seems like the acquire barrier must include that stage. (Actually, it can probably be _any_ stage as long as it matches `pWaitDstStageMask`?)

However, after applying @jherico's comments, the graphics acquire barrier has `srcStageMask = VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT` -- which _doesn't_ match the `pWaitDstStageMask`. So the queue ownership transfer barrier seems like it would not be blocked by the semaphore like it presumably needs to be.

Similar for the compute queue; semaphore wait is in `compute_submit_info.pWaitDstStageMask = VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT`, but the acquire barrier has `srcStageMask = VK_PIPELINE_STAGE_VERTEX_INPUT_BIT`.

**Comment by SaschaWillems** [comment]
Thanks for your input. I noticed that I still get validation layer errors (and one warning) with the fix mentioned by @jherico.

Will have to take a closer look at this.

**Comment by SaschaWillems** [comment]
@cforfang, @jherico: Can you chime in? I'm still struggling with that one validation layer message left. And none of the Vulkan implementations I have access too is having troubles with the current state of the sample, so I'm not even able to test if this change makes things better or worse.

**Comment by jherico** [comment]
I'll try to take a look tomorrow.  Do you have a commit ID that I can use to repro the validation message?

**Comment by SaschaWillems** [comment]
It's [this one](https://github.com/SaschaWillems/Vulkan-Samples/commit/e1069811d559e43193db56e078bed27071b4fa41) from my fork. I just did the changes that you commented some time ago. Thanks for checking :)

**Comment by jherico** [comment]
I've modified the code to resolve the validation error.  it runs successfully on my RTX 2080 without any validation errors.  The changes are here: https://github.com/SaschaWillems/Vulkan-Samples/pull/3
